Project Walking/Source code/app/src/main/java/com/teamapricot/projectwalking/LocationHandler.java

```diff
@@ -0,0 +1,153 @@
+
+/**
+ * A class for accessing location data.
+ */
+public class LocationHandler {
+    private final PermissionHandler permissionHandler;
+
+    private Position position = null;
+
+    private final LocationManager locationManager;
+    private LocationListener locationListener;
+
+    private final int updateInterval;
+
+    private final List<UpdateListener> updateListeners;
+
+    /**
+     * Constructor.
+     *
+     * @param activity The associated Activity
+     * @param interval Interval (in ms) between location updates
+     */
+    LocationHandler(AppCompatActivity activity, int interval) {
+        permissionHandler = new PermissionHandler(activity);
+        locationManager = (LocationManager) activity.getApplicationContext()
+                .getSystemService(AppCompatActivity.LOCATION_SERVICE);
+        updateListeners = new ArrayList<>();
+        updateInterval = interval;
+
+        if (permissionHandler.checkPermission(Manifest.permission.ACCESS_FINE_LOCATION)) {
+            initializeLocationHandler();
+        } else {
+            permissionHandler.requestPermissionAsync(Manifest.permission.ACCESS_FINE_LOCATION)
+                    .thenAccept(permissionGranted -> {
+                        if (permissionGranted) {
+                            initializeLocationHandler();
+                        } else {
+                            // TODO: Permission not granted, do something!
+                            System.out.println("permission not granted");
+                        }
+                    });
+        }
+    }
+
+    public Double getLatitude() {
+        if (position == null) {
+            return null;
+        }
+        return position.getLatitude();
+    }
+
+    public Double getLongitude() {
+        if (position == null) {
+            return null;
+        }
+        return position.getLongitude();
+    }
+
+    public Double getAccuracy() {
+        if (position == null) {
+            return null;
+        }
+        return position.getAccuracy();
+    }
+
+    /**
+     * Register an UpdateListener to get location updates.
+     *
+     * @param listener The listener to register
+     */
+    public void registerUpdateListener(UpdateListener listener) {
+        updateListeners.add(listener);
+    }
+
+    /**
+     * Unregister an UpdateListener.
+     *
+     * @param listener The listener to unregister
+     */
+    public void unregisterUpdateListener(UpdateListener listener) {
+        updateListeners.remove(listener);
+    }
+
+    private LocationListener createLocationListener() {
+        return location -> {
+            double latitude = location.getLatitude();
+            double longitude = location.getLongitude();
+            double accuracy = location.getAccuracy();
+
+            position = new Position(latitude, longitude, accuracy);
+
+            for (UpdateListener l : updateListeners) {
+                l.onPositionUpdated();
+            }
+        };
```

Comment:
Is there a reason why the user should use the `getLatitude`, `getLongitude` and `getAccuracy` methods in the `LocationHandler` to retrieve the updated values instead of having a parameter in your `UpdateListener`, like:

```java
public interface UpdateListener {
    void onPositionUpdated(Position position);
}
```

?

Reply:
In that case, I'd prefer

```java
public interface UpdateListener {
    void onPositionUpdated(double lat, double long, double acc);
}
```

to get rid of the need to call getters completely, instead of just switching out the object to call them on.

Reply:
Either is fine with me, I just think that the updated values should be provided in the `onPositionUpdated` method itself, since there are "edge cases" where the interval time could be so small that the values can change before the user manages to call all of the getter methods, right? This would lead to bad combinations of longitude, latitude and accuracy values. By providing the updated values as parameters in the method itself, we guarantee that the correct values are provided for that particular update. Not that it is a big deal in our case since the device itself will not move a lot between each location update, but I just feel like it is (don't hate me 😆) in general a "best practice" 😄 Let me know what you think.

Reply:
You definitely have a point there. I'll go ahead and make the Position class public again and use it, and replace the getters in LocationHandler with a getPosition.

Reply:
Oh, no hate whatsoever. :)